File: 747-MinCostClimbingStairs/747-MinCostClimbingStairs.py
# Last updated: 4/29/2026, 10:28:16 AM
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        def get_cost(i):
            if i < len(cost):
                return cost[i]
            else:
                return 0
        @cache
        def dp(i):
            if i >= len(cost):
                return 0
            return min(get_cost(i) + dp(i+1), get_cost(i+1) + dp(i+2))
        
        return dp(0)

File: 747-MinCostClimbingStairs/747-MinCostClimbingStairs.js
// Last updated: 4/29/2026, 10:28:11 AM
/**
 * @param {number[]} cost
 * @return {number}
 */
var minCostClimbingStairs = function(cost) {
    /*
    tc[i] = Math.min(cost[i] + tc[i+1], cost[i] + tc[i+2])
    return Math.min(tc[0], tc[1])
    */
    const map = {};
    const df = i => {
        if(map[i] !== undefined) {
            return map[i];
        }
        if(i >= cost.length) {
            return 0;
        }
        map[i] = Math.min(cost[i] + df(i+1), cost[i] + df(i+2));
        return map[i];
    }
    return Math.min(df(0), df(1));
};